RoboCleaner/HouseCleaningRobot.cs

```diff
@@ -1,6 +1,123 @@
+        {
+            return false;
+        }
+        visitedDirections.Add(direction);
+        return true;
+    }
+
+    /// <summary>
+    /// Whether the cell has been previously visited at all.
+    /// </summary>
+    /// <returns>True if the cell has been previously visited in any direction.</returns>
+    public bool PreviouslyVisited() => visitedDirections.Any();
+}
+
 public class HouseCleaningRobot
 {
-    // TODO - Implement
+    private readonly Cell[][] _grid;
+    private const Direction StartingDirection = Direction.Right;
+    private Position _startingPosition;
+
+    public HouseCleaningRobot(string[] input)
+    {
+        _grid = new Cell[input.Length][];
+        InitializeGrid(input);
+    }
+
+    private void InitializeGrid(string[] input)
+    {
+        for (int row = 0; row < _grid.Length; row++)
+        {
+            _grid[row] = new Cell[input[row].Length];
+            for (int column = 0; column < _grid[row].Length; column++)
+            {
+                _grid[row][column] = new Cell { IsObstacle = input[row][column] == CellTypes.Obstacle };
+                if (input[row][column] == CellTypes.StartingPoint)
+                {
+                    _startingPosition = new Position(row, column, StartingDirection);
+                }
+            }
+        }
+    }
+
+    public int CleanFloor()
+    {
+        Position currentPosition = _startingPosition;
+        int numCellsVisited = 1;
+        
+        while(Visit(currentPosition))
+        {
+            var movementResult = Move(currentPosition);
+            numCellsVisited += movementResult.NumberOfUniqueCellsVisited;
+            currentPosition = movementResult.NewPosition;
+        }
+        return numCellsVisited;
+    }
+
+    private bool Visit(Position position)
+    {
+        return _grid[position.Row][position.Column].Visit(position.Direction);
+    }
+
+    private MovementResult Move(Position currentPosition)
+    {
+        Position nextPosition = currentPosition.Direction switch
+        {
+            Direction.Left => currentPosition with { Column = currentPosition.Column - 1 },
+            Direction.Right => currentPosition with { Column = currentPosition.Column + 1 },
+            Direction.Up => currentPosition with { Row = currentPosition.Row - 1 },
+            Direction.Down => currentPosition with { Row = currentPosition.Row + 1 }
+        };
+
+        // Rotate
+        if (!IsValidPosition(nextPosition))
+        {
+            return new MovementResult(0, currentPosition with { Direction = Rotate(currentPosition.Direction) });
+        }
+
+        // Move
+        int cellsMoved = _grid[nextPosition.Row][nextPosition.Column].PreviouslyVisited() ? 0 : 1;
+        return new MovementResult(cellsMoved, nextPosition);
+    }
+
+    private bool IsValidPosition(Position position) =>
+        !(
+        position.Row < 0 || position.Row >= _grid.Length ||
+        position.Column < 0 || position.Column >= _grid.Length ||
```

Comment:
Should be _grid[position.Row].Length in order for this to work on non square grids.

Reply:
fixed.